lib/libc/newlib/libc-hooks.c

```diff
@@ -270,3 +268,147 @@ __weak int *__errno(void)
 {
 	return z_errno();
 }
+
+
+#if !defined(_RETARGETABLE_LOCKING) || \
+	CONFIG_NEWLIB_LIBC_DYNAMIC_LOCK_MEM_SIZE == 0
+
+static LIBC_DATA SYS_MUTEX_DEFINE(heap_mtx);
+
+void __malloc_lock(struct _reent *reent)
+{
+	sys_mutex_lock(&heap_mtx, K_FOREVER);
+}
+
+void __malloc_unlock(struct _reent *reent)
+{
+	sys_mutex_unlock(&heap_mtx);
+}
+
+#else
+#include <sys/mempool.h>
+
+#define NEWLIB_LOCK_POOL_SIZE WB_UP(CONFIG_NEWLIB_LIBC_DYNAMIC_LOCK_MEM_SIZE)
+
+#define NEWLIB_LOCK_FRAG_SIZE WB_UP(sizeof(void *) + sizeof(struct k_mutex))
+
+K_APPMEM_PARTITION_DEFINE(z_malloc_lock_partition);
+#define NEWLIB_LOCK_SECTION K_APP_DMEM_SECTION(z_malloc_lock_partition)
+
+
+SYS_MEM_POOL_DEFINE(z_nl_lock_pool, NULL,
+		NEWLIB_LOCK_FRAG_SIZE, NEWLIB_LOCK_POOL_SIZE,
+		1, sizeof(void *), NEWLIB_LOCK_SECTION);
+
+static int newlib_lock_prepare(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	sys_mem_pool_init(&z_nl_lock_pool);
+
+	return 0;
+}
+
+SYS_INIT(newlib_lock_prepare, PRE_KERNEL_1,
+	CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_sinit_recursive_mutex);
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_sfp_recursive_mutex);
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_atexit_recursive_mutex);
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_malloc_recursive_mutex);
+static LIBC_DATA SYS_MUTEX_DEFINE(nl_env_recursive_mutex);
+static LIBC_DATA SYS_SEM_DEFINE(nl_at_quick_exit_sem, 1, 1);
+static LIBC_DATA SYS_SEM_DEFINE(nl_tz_sem, 1, 1);
+static LIBC_DATA SYS_SEM_DEFINE(nl_dd_hash_sem, 1, 1);
+static LIBC_DATA SYS_SEM_DEFINE(nl_arc4random_sem, 1, 1);
+
+struct __lock {
+	void *pSemOrMtx;
+};
+
+struct __lock __lock___sinit_recursive_mutex = {&nl_sinit_recursive_mutex};
+struct __lock __lock___sfp_recursive_mutex = {&nl_sfp_recursive_mutex};
+struct __lock __lock___atexit_recursive_mutex = {&nl_atexit_recursive_mutex};
+struct __lock __lock___malloc_recursive_mutex = {&nl_malloc_recursive_mutex};
+struct __lock __lock___env_recursive_mutex = {&nl_env_recursive_mutex};
+
+struct __lock __lock___at_quick_exit_mutex = {&nl_at_quick_exit_sem};
+struct __lock __lock___tz_mutex = {&nl_tz_sem};
+struct __lock __lock___dd_hash_mutex = {&nl_dd_hash_sem};
+struct __lock __lock___arc4random_mutex = {&nl_arc4random_sem};
+
+
+void __retarget_lock_init(_LOCK_T *lock)
+{
+	lock = sys_mem_pool_alloc(&z_nl_lock_pool, NEWLIB_LOCK_FRAG_SIZE);
+
+	__ASSERT(lock, "failed to allocate memory for newlib lock");
+
+	(*lock)->pSemOrMtx = (void *) &((char *)lock)[sizeof(void *)];
```

Comment:
Same comments here as in the other init function.

Reply:
Changed that